tokio-util/src/net/unix/mod.rs

```diff
@@ -0,0 +1,108 @@
+//! Unix domain socket helpers.
+
+use std::future::Future;
+use std::io::Result;
+use std::pin::Pin;
+
+/// Listening address.
+#[derive(Debug)]
+pub enum ListenAddr {
+    /// Socket address.
+    SocketAddr(std::net::SocketAddr),
+    /// Unix socket.
+    UnixSocket(tokio::net::unix::SocketAddr),
+}
+
+impl From<std::net::SocketAddr> for ListenAddr {
+    fn from(addr: std::net::SocketAddr) -> Self {
+        Self::SocketAddr(addr)
+    }
+}
+
+impl From<tokio::net::unix::SocketAddr> for ListenAddr {
+    fn from(addr: tokio::net::unix::SocketAddr) -> Self {
+        Self::UnixSocket(addr)
+    }
+}
+
+/// A trait for a listener: `TcpListener` and `UnixListener`.
+pub trait Listener: Send + Unpin {
+    /// Accepts a new incoming connection from this listener.
+    fn accept<'a>(
+        &'a self,
+    ) -> Pin<
+        Box<
+            dyn Future<
+                    Output = Result<(Box<dyn AsyncReadWrite + Send + Unpin + 'static>, ListenAddr)>,
+                > + Send
+                + 'a,
+        >,
+    >;
+
+    /// Returns the local address that this listener is bound to.
+    fn local_addr(&self) -> Result<ListenAddr>;
```

Comment:
Also seems like this should be an associated type rather than return an enum with the two types hard-coded. Otherwise the trait is designed to _only_ allow `TcpListener` and `UnixListener`, and people would not be able to implement it for their own listeners, e.g. maybe there's a listener for TLS streams.

Reply:
My idea was to make it **only** for TcpListener and UnixListener :grin: it didn't even crossed my mind to support other things.

I don't mind working on making this more generic but I need to update the title of this issue then.